src/test/dart/devtools/devtools.test.ts

```diff
@@ -50,5 +50,32 @@ describe("DevTools", async () => {
 		const openedUri = openBrowserCommand.args[0][0] as string; // First invocation, first arg.
 		assertDevToolsUriWithVmService(openedUri, debuggerUrisEvent.vmServiceUri);
 	});
+
+	it("opens with the correct client VM service in the URI", async () => {
+		const debugSession = startFakeDebugSession();
+		const debuggerUrisEvent = {
+			clientVmServiceUri: "ws://fake-client:456/ws",
+			vmServiceUri: "ws://fake-host:123/ws",
+		};
+		extApi.debugCommands.handleDebugSessionCustomEvent({
+			body: debuggerUrisEvent,
+			event: "dart.debuggerUris",
+			session: debugSession,
+		});
+
+		// Stub out openInBrowser so we don't really open and can capture the arguments.
+		const openBrowserCommand = sb.stub(extApi.envUtils, "openInBrowser").resolves();
+
+		// Trigger opening DevTools.
+		await vs.commands.executeCommand("dart.openDevTools.external");
+
+		// Verify an attempt was made to open the correct URI.
+		assert.equal(openBrowserCommand.calledOnce, true);
+		const openedUri = openBrowserCommand.args[0][0] as string; // First invocation, first arg.
+		assert.ok(
+			openedUri.includes(debuggerUrisEvent.clientVmServiceUri),
+			`URI did not contain expected VM Service!\nOpened: ${openedUri}\nExpected to contain: ${debuggerUrisEvent.clientVmServiceUri}`,
+		);
+	});
```

Comment:
I think this should use `assertDevToolsUriWithVmService`. Is there an issue with the encoding when you used it? If so, we should understand whether the encoding is incorrect, or if the other URL is just over-encoded or something and try and make them consistent.

Reply:
Oops sorry, I got confused and thought this was simulating the port forwarding from `exposeUrl`.